### PR TITLE
Allow setting of a demodulized class name when using STI

### PR DESCRIPTION
**Pull request by agrobbin**
If your STI class looks like this:

``` ruby
class Company < ActiveRecord::Base
  self.store_full_sti_class = false

  class GoodCo < Company
  end

  class BadCo < Company
  end
end
```

The expectation (which is valid) is that the `type` in the database is saved as `GoodCo` or `BadCo`. However, another expectation should be that setting `type` to `GoodCo` would correctly instantiate the object as a `Company::GoodCo`. That second expectation is what this should fix.

The case that brought this to my attention is when I'm dynamically defining the `type` as part of the hash of attributes. Before this patch, an exception is raised:

```
ActiveRecord::SubclassNotFound: Invalid single-table inheritance type: SpecialCo is not a subclass of Company
```

### CODE REVIEW

File: activerecord/test/cases/inheritance_test.rb [hunk 1]
```diff
@@ -208,6 +208,16 @@ def test_new_with_complex_inheritance
     assert_nothing_raised { Client.new(type: 'VerySpecialClient') }
   end
 
+  def test_new_with_store_full_sti_class_option
+    old = ActiveRecord::Base.store_full_sti_class
+    ActiveRecord::Base.store_full_sti_class = false
+    Company::SpecialCo # force autoloading
+    item = Company.new(:type => 'SpecialCo')
+    assert_equal Company::SpecialCo, item.class
```
Comment (senny) [inline activerecord/test/cases/inheritance_test.rb]: We have `assert_instance_of`

File: activerecord/test/cases/inheritance_test.rb [hunk 2]
```diff
@@ -208,6 +208,16 @@ def test_new_with_complex_inheritance
     assert_nothing_raised { Client.new(type: 'VerySpecialClient') }
   end
 
+  def test_new_with_store_full_sti_class_option
```
Comment (senny) [inline activerecord/test/cases/inheritance_test.rb]: The current test name `test_new_with_store_full_sti_class_option` makes me think it uses a `true` value for the option. Can we reword it to reflect that it's disabling the option?

**Comment by agrobbin** [comment]
All set @senny!

File: activerecord/test/cases/inheritance_test.rb [hunk 3]
```diff
@@ -208,6 +208,16 @@ def test_new_with_complex_inheritance
     assert_nothing_raised { Client.new(type: 'VerySpecialClient') }
   end
 
+  def test_new_without_storing_full_sti_class
+    old = ActiveRecord::Base.store_full_sti_class
```
Comment (senny) [inline activerecord/test/cases/inheritance_test.rb]: I pushed a little refactoring containing a helper to assign the configuration (https://github.com/rails/rails/commit/d27974d3fb38918d323dcce3ae439ef68b369b72). Could you update this test case to use the helper as well?

File: activerecord/test/cases/inheritance_test.rb [hunk 4]
```diff
@@ -208,6 +208,16 @@ def test_new_with_complex_inheritance
...
+    ActiveRecord::Base.store_full_sti_class = false
+    Company::SpecialCo # force autoloading
+    item = Company.new(:type => 'SpecialCo')
```
Comment (senny) [inline activerecord/test/cases/inheritance_test.rb]: let's use the new hash syntax here.

File: activerecord/test/models/company.rb [hunk 5]
```diff
@@ -26,6 +26,9 @@ def arbitrary_method
   def private_method
     "I am Jack's innermost fears and aspirations"
   end
+
+  class SpecialCo < Company
+  end
```
Comment (senny) [inline activerecord/test/models/company.rb]: Do we need a new model or can we use `Namespaced::Company` to reproduce the error as well?

```ruby
Namespaced::Company.new(type: "Firm")
```

**Comment by senny** [comment]
@agrobbin sorry for the delay, I've added a few minor comments. Could you add an entry to the changelog as well?

Other than that the patch looks good :+1:

File: activerecord/test/models/company.rb (hunk 5 above)
Comment (agrobbin) [inline activerecord/test/models/company.rb]: When I attempt to replace my model with that, I receive this error:

```
InheritanceTest#test_new_without_storing_full_sti_class:
ActiveRecord::SubclassNotFound: Invalid single-table inheritance type: Firm is not a subclass of Namespaced::Company
```

I'm not sure exactly what causes that failure. Any thoughts?

Comment (agrobbin) [inline activerecord/test/models/company.rb]: Perhaps there is another class `Firm` that is outside of the `Namespaced` module?

Comment (agrobbin) [inline activerecord/test/models/company.rb]: Ah, indeed [there is](https://github.com/rails/rails/blob/db0e649aba7135133154a35549219cffe968eaf6/activerecord/test/models/company.rb#L43)! That causes `find_sti_class` to blow up.

Comment (senny) [inline activerecord/test/models/company.rb]: @agrobbin seems like a bug :sweat:. You could try one of the other subclasses in the `Namespaced` module to get this fix out the door. Would you be interested to look at the issue with `find_sti_class` in combination with `Firm`?

Comment (agrobbin) [inline activerecord/test/models/company.rb]: @senny unfortunately the only other subclass within the `Namespaced` module is `Client`, which [also exists outside of the module](https://github.com/rails/rails/blob/db0e649aba7135133154a35549219cffe968eaf6/activerecord/test/models/company.rb#L116).

Looking into the other issue very quickly, it appears that `find_sti_class` is working correctly:

```ruby
subclass = find_sti_class(subclass_name) # => Namespaced::Firm(id: integer, type: string, firm_id: integer, firm_name: string, name: string, client_of: integer, rating: integer, account_id: integer, description: string)
```

However, `descendants` is returning an empty array for some reason. I know very little about `ActiveSupport::DescendantsTracker`, so I'm not quite sure where to start investigating why it would be empty, particularly after forcing autoloading of `Namespaced::Firm`.

**Comment by senny** [comment]
@agrobbin thanks for your work :yellow_heart: 

Everything seemed to work. The error you got when using `Namespaced::Company` was because the class `Namespaced::Firm` extends `::Company` (top level) and not the namespaced one. This is fine but the error message only stated `Firm`, which was a bit confusing. I pushed f7881d3e2b87016c6788ceceda4c849ae7a0a009 to include the full class name in the error message.

**Comment by agrobbin** [comment]
Thanks @senny!
